File: binary_tree/traverse_a_tree/binary_tree_inorder_traversal/binary_tree_inorder_traversal_recursive.java
/**
 * Author: Nathan Bockisch
 * Date: October 13, 2021
 **/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> values = new ArrayList<Integer>();
        
        if (root == null) return values;
        
        values.addAll(inorderTraversal(root.left));
        values.add(root.val);
        values.addAll(inorderTraversal(root.right));
        
        return values;
    }
}

File: recursion_ii/recursion_to_iteration/binary_tree_inorder_traversal/binary_tree_inorder_traversal_recursive.py
'''
Author: Nathan Bockisch
Date: November 5, 2021
'''
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:   
        def traverse(cur: TreeNode, nodes: list) -> List[int]:
            if (cur):
                traverse(cur.left, nodes)
                nodes.append(cur.val)
                traverse(cur.right, nodes)
            
            return nodes
        
        return traverse(root, [])
